src/ZkFold/Base/Data/Vector.hs

```diff
@@ -5,119 +5,109 @@
-import           Data.Functor.Rep                 (Representable (..), collectRep, distributeRep)
-import qualified Data.List                        as List
-import           Data.List.Split                  (chunksOf)
+import           Data.Foldable                    (fold)
+import           Data.Functor.Rep                 (Representable (..), collectRep, distributeRep, mzipRep, pureRep)
 import           Data.These                       (These (..))
+import qualified Data.Vector                      as V
+import           Data.Vector.Binary               ()
+import qualified Data.Vector.Split                as V
 import           Data.Zip                         (Semialign (..), Zip (..))
 import           GHC.Generics                     (Generic)
 import           GHC.IsList                       (IsList (..))
-import           Prelude                          hiding (drop, head, length, mod, replicate, sum, tail, take, zip,
-                                                   zipWith, (*))
-import qualified Prelude                          as P
+import           Prelude                          hiding (concat, drop, head, length, mod, replicate, sum, tail, take,
+                                                   zip, zipWith, (*))
 import           System.Random                    (Random (..))
 import           Test.QuickCheck                  (Arbitrary (..))
 
 import           ZkFold.Base.Algebra.Basic.Class
 import           ZkFold.Base.Algebra.Basic.Field
 import           ZkFold.Base.Algebra.Basic.Number
 import           ZkFold.Base.Data.ByteString      (Binary (..))
-import qualified ZkFold.Prelude                   as ZP
-import           ZkFold.Prelude                   (length, replicate)
+import           ZkFold.Prelude                   (length)
 
-newtype Vector (size :: Natural) a = Vector [a]
+newtype Vector (size :: Natural) a = Vector {toV :: V.Vector a}
     deriving (Show, Eq, Functor, Foldable, Traversable, Generic, NFData, Ord)
 
+-- helper
+knownNat :: forall size n . (KnownNat size, Integral n) => n
+knownNat = fromIntegral (value @size)
+
 instance KnownNat size => Representable (Vector size) where
   type Rep (Vector size) = Zp size
-  index (Vector v) ix = v Prelude.!! (fromIntegral (fromZp ix))
-  tabulate f = Vector [f (toZp ix) | ix <- [0 .. fromIntegral (value @size) Prelude.- 1]]
+  index (Vector v) ix = v V.! (fromIntegral (fromZp ix))
+  tabulate f = Vector (V.generate (knownNat @size) (f . fromIntegral))
 
 instance KnownNat size => Distributive (Vector size) where
   distribute = distributeRep
   collect = collectRep
 
+
+vtoVector :: forall size a . KnownNat size => V.Vector a -> Maybe (Vector size a)
+vtoVector as
+  | V.length as == knownNat @size = Just $ Vector as
+  | otherwise                     = Nothing
+
 instance IsList (Vector n a) where
     type Item (Vector n a) = a
     toList = fromVector
     fromList = unsafeToVector
 
-parFmap :: (a -> b) -> Vector size a -> Vector size b
```

Comment:
Btw why it doesn't error out? I thought we were using it somewhere

Also maybe we can add some analog of it back?

Reply:
It’s not used anywhere